### PR TITLE
QuartzHostedService now has jobs start after application startup

### PR DESCRIPTION
**Pull request by Timovzl**
Implements #1432.

QuartzHostedService now has jobs start after application startup completes, unless QuartzHostedServiceOptions.AwaitApplicationStarted=false.

### CODE REVIEW

**Review by lahma — APPROVED** [review]
Looks good to me. If @andrewlock does not have improvement ideas I will merge this in a few days. Thanks for this splendid contribution, kudos for the test coverage!

**Comment by Timovzl** [comment]
@lahma I wanted to check with you that ISchedulerFactory.GetScheduler() does _not_ cause job exection to start. Can you confirm that? Since we’re still calling that immediately (which seems good for perhaps ensuring connectivity to the remote).

**Comment by Timovzl** [comment]
I’m also now considering the removal of lines 36-42 of QuartzHostedService, since I just realized that the task never completes synchronously (unless cancellation _just_ occurred, in which case we do not care about the exception).

Seems like I followed BackgroundService’s pattern needlessly in this case.

**Review by Timovzl — COMMENTED** [review]
TO DO: Remove QuartzHostedService lines 36-42, as the task does not complete synchronously?

**Comment by lahma** [comment]
> I wanted to check with you that ISchedulerFactory.GetScheduler() does _not_ cause job exection to start. Can you confirm that? Since we’re still calling that immediately (which seems good for perhaps ensuring connectivity to the remote).

The scheduler starts when `Start()` is called, it can be "no-op" when it's configured with `ZeroSizeThreadPool`, but the `Start` should be the key here. There are some funky code paths which require `GetScheduler` to be Task, can't remember where the rabbit hole went... Generally it should always be non-async in real like (until proven otherwise).

> TO DO: Remove QuartzHostedService lines 36-42, as the task does not complete synchronously?

I'm not sure I quite follow.

**Comment by Timovzl** [comment]
Those lines contain some code that await the task if it is already completed, so as to propagate any exceptions in that case.

However, since the Task starts with a delay, it practically never completes synchronously, so that code is redundant.

**Comment by lahma** [comment]
> Those lines contain some code that await the task if it is already completed, so as to propagate any exceptions in that case.
> 
> However, since the Task starts with a delay, it practically never completes synchronously, so that code is redundant.

Sounds reasonable, I trust your judgement here.

**Comment by Timovzl** [comment]
Actually, scratch that. Changed my mind. It’s more correct this way in the case where ApplicationStarted is already triggered (tests?) and it’s better to follow the standard pattern. Leaving it as-is.

File: src/Quartz.Extensions.Hosting/QuartzHostedService.cs [hunk 1]
```diff
@@ -1,44 +1,94 @@
...
+                // Follow the pattern from BackgroundService.StartAsync: https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs
+
+                var deferredSchedulerStartTask = AwaitStartupCompletionAndStartSchedulerAsync(cancellationToken);
+
+                // If the task completed synchronously, await it in order to bubble potential cancellation/failure to the caller
+                // Otherwise, return, allowing application startup to complete
+                if (deferredSchedulerStartTask.IsCompleted)
+                {
+                    await deferredSchedulerStartTask.ConfigureAwait(false);
+                }
```
Comment (andrewlock) [inline src/Quartz.Extensions.Hosting/QuartzHostedService.cs:42]: I don't think this pattern is quite correct 🤔 If `!deferredSchedulerStartTask.IsCompleted`, I think you need to store that task in a field and await it in `StopAsync`, similar to [the way `BackgroundService` works](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting.Abstractions/src/BackgroundService.cs)?

Comment (Timovzl) [reply src/Quartz.Extensions.Hosting/QuartzHostedService.cs:42]: I think you're right. I'll make a change.

(For the sake of completeness, note that this is quite niche. It is only for one scenario: when the application is stopped _while our deferred startup task has not yet completed_. It achieves only one thing: whilst we are cancelling, the await in `StopAsync` helps shutdown wait until our cancellation completes. For `BackgroundService` it is much more relevant, as its execute task is expected to keep running during the application lifetime.)

File: src/Quartz.Tests.Unit/QuartzHostedServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+
+            using var startupCts = new CancellationTokenSource();
+
+            await quartzHostedService.StartAsync(startupCts.Token);
+
+            Assert.NotNull(schedulerFactory.LastCreatedScheduler);
+            Assert.AreEqual(shouldSchedulerBeStartedImmediately, schedulerFactory.LastCreatedScheduler.IsStarted);
+
+            appliationLifetime.SetStarted();
+            await Task.Delay(20); // Give the hosted service some time to respond to the ApplicationStarted token
```
[the remark this answers is not in the document]
Comment (Timovzl) [reply src/Quartz.Tests.Unit/QuartzHostedServiceTests.cs:391]: Agreed.

**Comment by Timovzl** [comment]
@andrewlock I have implemented your suggestions.

**Comment by Timovzl** [comment]
@lahma I see that `TestNoConcurrentExecOnSameJob` has failed. I find it hard to imagine that this was caused by the change. Could it be that this is a sensitive test that can sometimes fail due to timing or concurrency issues? Could you rerun the check to see if it goes away?

**Comment by lahma** [comment]
Don't worry about it, some tests might have timing issues, single failing agent doesn't rock the boat - I'm rerunning them though.

**Comment by lahma** [comment]
Great work, thank you!

**Pull request merged**